File: cversion/machine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "machine.h"
#include "cpu.h"

Cpu initial_cpu_state()
{
	Cpu cpu;
	cpu.a = 0x01;
	cpu.b = 0x00;
	cpu.c = 0x13;
	cpu.d = 0x00;
	cpu.e = 0xD8;
	cpu.f = 0xB0;
	cpu.h = 0x01;
	cpu.l = 0x4D;
	cpu.pc = 0x0100;
	cpu.sp = 0xFFFE;

	return cpu;
}

void initial_memory_state(Memory* memory)
{
	memory->ram[0xFF05] = 0x00;
	memory->ram[0xFF06] = 0x00;
	memory->ram[0xFF07] = 0x00;
	memory->ram[0xFF10] = 0x80;
	memory->ram[0xFF11] = 0xBF;
	memory->ram[0xFF12] = 0xF3;
	memory->ram[0xFF14] = 0xBF;
	memory->ram[0xFF16] = 0x3F;
	memory->ram[0xFF17] = 0x00;
	memory->ram[0xFF19] = 0xBF;
	memory->ram[0xFF1A] = 0x7F;
	memory->ram[0xFF1B] = 0xFF;
	memory->ram[0xFF1C] = 0x9F;
	memory->ram[0xFF1E] = 0xBF;
	memory->ram[0xFF20] = 0xFF;
	memory->ram[0xFF21] = 0x00;
	memory->ram[0xFF22] = 0x00;
	memory->ram[0xFF23] = 0xBF;
	memory->ram[0xFF24] = 0x77;
	memory->ram[0xFF25] = 0xF3;
	memory->ram[0xFF26] = 0xF1;
	memory->ram[0xFF40] = 0x91;
	memory->ram[0xFF42] = 0x00;
	memory->ram[0xFF43] = 0x00;
	memory->ram[0xFF45] = 0x00;
	memory->ram[0xFF47] = 0xFC;
	memory->ram[0xFF48] = 0xFF;
	memory->ram[0xFF49] = 0xFF;
	memory->ram[0xFF4A] = 0x00;
	memory->ram[0xFF4B] = 0x00;
	memory->ram[0xFFFF] = 0x00;
}

int init_machine(Machine* machine, Cpu* cpu, Memory* memory)
{
	machine->cpu = cpu;
	machine->memory = memory;

	*(machine->cpu) = initial_cpu_state();
	initial_memory_state(machine->memory);

	machine->ime = 0;
	machine->vblank_counter = 0;
	machine->hblank_counter = 0;
	machine->hblank_mode3_counter = 80;
	machine->hblank_mode0_counter = 80 + 172;
	machine->current_scanline = 153;
	machine->div_counter = 0;

	machine->display = malloc(LCD_WIDTH * LCD_HEIGHT);
	if(machine->display == 0)
		return -1;

	memset(machine->display, 0, LCD_WIDTH * LCD_HEIGHT);

	return 0;
}

void free_machine(Machine* machine)
{
	free(machine->display);
}

void instruction_error_handler(void* data)
{
	fprintf(stderr, "INVALID CPU INSTRUCTION\n");
	abort();
}

inline static void render_scanline(Machine* machine)
{
	int scanline = machine->current_scanline;

	int scx = machine->memory->ram[0xFF43];
	int scy = machine->memory->ram[0xFF42];

	int lcdc = machine->memory->ram[0xFF40];

	/*
	printf("%3d LCDC: %d%d%d%d%d%d%d%d\n", scanline,
			(lcdc&(1<<7))>0,
			(lcdc&(1<<6))>0,
			(lcdc&(1<<5))>0,
			(lcdc&(1<<4))>0,
			(lcdc&(1<<3))>0,
			(lcdc&(1<<2))>0,
			(lcdc&(1<<1))>0,
			(lcdc&(1<<0))>0);
	*/

	if((lcdc & (1 << 7)) == 0)
	{
		int x;
		for(x = 0; x < LCD_WIDTH; x++)
			machine->display[scanline * LCD_WIDTH + x] = 0;
		return;
	}

	int bgmap_start_addr = (lcdc & (1 << 3)) ? 0x9C00 : 0x9800;

	int bgtiles_start_addr = (lcdc & (1 << 4)) ? 0x8000 : 0x9000;

	int x;
	int y = scanline;
	for(x = 0; x < LCD_WIDTH; x++)
	{
		int xp = (x + scx) % 256;
		int yp = (y + scy) % 256;
		int xrow = xp / 8;
		int yrow = yp / 8;
		int tile_num = yrow * 32 + xrow;
		int tile_index = machine->memory->ram[tile_num + bgmap_start_addr];
		int tile_start_mem = bgtiles_start_addr + (16 * tile_index);
		int xoff = 7 - (xp % 8);
		int yoff = yp % 8;
		int hi_byte = tile_start_mem + (yoff * 2);
		int lo_byte = tile_start_mem + (yoff * 2) + 1;
		int hi_byte_value = machine->memory->ram[hi_byte];
		int lo_byte_value = machine->memory->ram[lo_byte];
		int color = 2*((lo_byte_value & (1 << xoff)) > 0) + ((hi_byte_value & (1 << xoff)) > 0);
		machine->display[y * LCD_WIDTH + x] = color;
	}
}

/* returns 1 if a vblank occured
 * otherwise returns 0 */
inline static int irq_update(Machine* machine, int cycles)
{
	int vblank_flag = 0;

	machine->vblank_counter -= cycles;
	machine->hblank_counter -= cycles;
	machine->hblank_mode3_counter -= cycles;
	machine->hblank_mode0_counter -= cycles;
	machine->div_counter -= cycles;

	if(machine->hblank_counter <= 0)
	{
		machine->hblank_counter += HBLANK_PERIOD;
		machine->memory->ram[0xFF44] = (machine->memory->ram[0xFF44] + 1) % 154;
		machine->current_scanline = (machine->current_scanline + 1) % 154;
		if (machine->current_scanline < 144)
		{
			machine->memory->ram[0xFF41] &= ~(1 << 0);
			machine->memory->ram[0xFF41] |= (1 << 1);
			if(machine->memory->ram[0xFF44] == machine->memory->ram[0xFF45])
				machine->memory->ram[0xFF41] |= (1 << 2);
			else
				machine->memory->ram[0xFF41] &= ~(1 << 2);

			if(machine->memory->ram[0xFF41] & (1 << 5) ||
					(machine->memory->ram[0xFF41] & (1 << 6) &&
					 machine->memory->ram[0xFF44] ==
					 machine->memory->ram[0xFF45]))
				machine->memory->ram[0xFF0F] |= (1 << 1);

			render_scanline(machine);
		}
	}

	if(machine->hblank_mode3_counter <= 0)
	{
		machine->hblank_mode3_counter += HBLANK_PERIOD;
		if(machine->current_scanline < 144)
		{
			machine->memory->ram[0xFF41] |= (1 << 0);
			machine->memory->ram[0xFF41] |= (1 << 1);
		}
	}

	if(machine->hblank_mode0_counter <= 0)
	{
		machine->hblank_mode0_counter += HBLANK_PERIOD;
		if(machine->current_scanline < 144)
		{
			machine->memory->ram[0xFF41] &= ~(1 << 0);
			machine->memory->ram[0xFF41] &= ~(1 << 1);
			if(machine->memory->ram[0xFF41] & (1 << 3))
				machine->memory->ram[0xFF0F] |= (1 << 1);
		}
	}

	if(machine->vblank_counter <= 0)
	{
		vblank_flag = 1;
		machine->vblank_counter += VBLANK_PERIOD;
		machine->memory->ram[0xFF41] |= (1 << 0);
		machine->memory->ram[0xFF41] &= ~(1 << 1);
		if(machine->memory->ram[0xFF41] & (1 << 4))
			machine->memory->ram[0xFF0F] |= (1 << 1);
		machine->memory->ram[0xFF0F] |= (1 << 0);
	}

	if(machine->div_counter <= 0)
	{
		machine->div_counter += DIV_PERIOD;
		machine->memory->ram[0xFF04] += 1;
	}

	int flagsIF = machine->memory->ram[0xFF0F];
	int flagsIE = machine->memory->ram[0xFFFF];
	if(machine->ime && (flagsIF & flagsIE))
	{
		int i = 0;
		while(1)
		{
			if (flagsIF & flagsIE & (1 << i))
				break;
			i++;
		}
		machine->ime = 0;
		machine->memory->ram[0xFF0F] &= ~(1 << i);
		u16 oldPC = machine->cpu->pc;
		u8 hiPC, loPC;
		split_word_16(oldPC, &hiPC, &loPC);
		u16 oldSP = machine->cpu->sp;
		machine->memory->ram[oldSP-1] = hiPC;
		machine->memory->ram[oldSP-2] = loPC;
		machine->cpu->sp -= 2;
		switch(i)
		{
			case 0: machine->cpu->pc = 0x0040; break;
			case 1: machine->cpu->pc = 0x0048; break;
			case 2: machine->cpu->pc = 0x0050; break;
			case 3: machine->cpu->pc = 0x0058; break;
			case 4: machine->cpu->pc = 0x0060; break;
			default: assert(0);
		}
	}

	return vblank_flag;
}

/* returns 1 if a vblank occured
 * otherwise returns 0 */
inline int update_machine(Machine* machine, u8* breakpoint_table)
{
/*
	if(breakpoint_table)
	{
		int bp = query_breakpoint(breakpoint_table, machine->cpu->pc);
		if(bp)
		{
*/

/*
			printf("\
%04X: %02X\n\
af = %02X%02X  z = %d  %04X: %02X\n\
bc = %02X%02X  n = %d  %04X: %02X\n\
de = %02X%02X  h = %d  %04X: %02X\n\
hl = %02X%02X  c = %d  %04X: %02X\n\
sp = %04X  %cI     %04X: %02X\n\
pc = %04X  ROM01  %04X: %02X\n\
                  %04X: %02X\n\
IE:%d%d%d%d%d%d%d%d       %04X: %02X\n\
SCX:%02x SCY:%02x     %04X: %02X\n",
				(int)machine->cpu->pc, (int)read_memory(machine->memory, machine->cpu->pc),
				(int)machine->cpu->a, (int)machine->cpu->f,
				(machine->cpu->f & (1 << 7)) > 0,
				machine->cpu->sp-4,
				read_memory(machine->memory, machine->cpu->sp-4),
				(int)machine->cpu->b, (int)machine->cpu->c,
				(machine->cpu->f & (1 << 6)) > 0,
				machine->cpu->sp-3,
				read_memory(machine->memory, machine->cpu->sp-3),
				(int)machine->cpu->d, (int)machine->cpu->e,
				(machine->cpu->f & (1 << 5)) > 0,
				machine->cpu->sp-2,
				read_memory(machine->memory, machine->cpu->sp-2),
				(int)machine->cpu->h, (int)machine->cpu->l,
				(machine->cpu->f & (1 << 4)) > 0,
				machine->cpu->sp-1,
				read_memory(machine->memory, machine->cpu->sp-1),
				(int)machine->cpu->sp,
				machine->ime ? 'E' : 'D',
				machine->cpu->sp,
				read_memory(machine->memory, machine->cpu->sp),
				(int)machine->cpu->pc,
				machine->cpu->sp+1,
				read_memory(machine->memory, machine->cpu->sp+1),
				machine->cpu->sp+2,
				read_memory(machine->memory, machine->cpu->sp+2),
				(machine->memory->ram[0xFFFF] & (1 << 7)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 6)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 5)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 4)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 4)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 2)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 1)) > 0,
				(machine->memory->ram[0xFFFF] & (1 << 0)) > 0,
				machine->cpu->sp+3,
				read_memory(machine->memory, machine->cpu->sp+3),
				(int)machine->memory->ram[0xFF43],
				(int)machine->memory->ram[0xFF42],
				machine->cpu->sp+4,
				read_memory(machine->memory, machine->cpu->sp+4));
*/


/*
			char c;
			scanf("%c", &c);
		}
	}
*/

	int cycle_count = step_instruction(machine->cpu,
			machine->memory,
			machine,
			instruction_error_handler, 0);
	return irq_update(machine, cycle_count);
}

void update_machine_vblank(Machine* machine, u8* breakpoint_table)
{
	while(update_machine(machine, breakpoint_table) == 0);
}

File: src/Machine.hs
-- OmegaGB Copyright 2007 Bit Connor
-- This program is distributed under the terms of the GNU General Public License
-----------------------------------------------------------------------------
-- |
-- Module      :  Machine
-- Copyright   :  (c) Bit Connor 2007 <bit@mutantlemon.com>
-- License     :  GPL
-- Maintainer  :  bit@mutantlemon.com
-- Stability   :  in-progress
--
-- OmegaGB
-- Game Boy Emulator
--
-- This module does emulation of a lot of the Game Boy's hardware, including
-- interrupts and graphics rendering. It's still incomplete.
-- 
-- The complete state of all of the GameBoy is represented as the tuple:
-- ((RegisterStates, Memory), IrqStates)
--
-----------------------------------------------------------------------------

{-# LANGUAGE GADTs #-}

module Machine where
--import Prerequisites

import Data.Array.IArray
import Data.Word
import Data.Int
import Data.Bits
import Data.Maybe
import Data.List
import Control.Monad.State

import WordUtil
import Cpu
import CpuExecution
import Memory
import Display
import Joypad

import Debug.Trace

--cpuClockFrequency :: Double
--cpuClockFrequency = 4194304

--horizSync :: Double
--horizSync = 9198000

--vertSync :: Double
--vertSync = 59.73

--vblankPeriod = cpuClockFrequency / vertSync

--scanlinePeriod = vblankPeriod / 153

-- Strict field used to contain states
data RegisterStates = RegisterStates
       {-# UNPACK #-} !Word8  -- A
       {-# UNPACK #-} !Word8  -- B
       {-# UNPACK #-} !Word8  -- C
       {-# UNPACK #-} !Word8  -- D
       {-# UNPACK #-} !Word8  -- E
       {-# UNPACK #-} !Word8  -- F
       {-# UNPACK #-} !Word8  -- H
       {-# UNPACK #-} !Word8  -- L
       {-# UNPACK #-} !Word16 -- PC
       {-# UNPACK #-} !Word16 -- SP

getRegState :: RegisterStates -> M_Register -> Word8
getRegState rs r =
  let RegisterStates a b c d e f h l _ _ = rs in
  case r of
    M_A -> a
    M_B -> b
    M_C -> c
    M_D -> d
    M_E -> e
    M_F -> f
    M_H -> h
    M_L -> l

setRegState :: RegisterStates -> M_Register -> Word8 -> RegisterStates
setRegState rs r n =
  let RegisterStates a b c d e f h l pc sp = rs in
  case r of
    M_A -> RegisterStates n b c d e f h l pc sp
    M_B -> RegisterStates a n c d e f h l pc sp
    M_C -> RegisterStates a b n d e f h l pc sp
    M_D -> RegisterStates a b c n e f h l pc sp
    M_E -> RegisterStates a b c d n f h l pc sp
    M_F -> RegisterStates a b c d e (n.&.0xF0) h l pc sp
    M_H -> RegisterStates a b c d e f n l pc sp
    M_L -> RegisterStates a b c d e f h n pc sp

getReg2State :: RegisterStates -> M_Register2 -> Word16
getReg2State rs r2 =
  let RegisterStates a b c d e f h l pc sp = rs in
  case r2 of
    M_AF -> joinWord16 a f
    M_BC -> joinWord16 b c
    M_DE -> joinWord16 d e
    M_HL -> joinWord16 h l
    M_PC -> pc
    M_SP -> sp

setReg2State :: RegisterStates -> M_Register2 -> Word16 -> RegisterStates
setReg2State rs r2 nn =
  let RegisterStates a b c d e f h l pc sp = rs
      (hi, lo) = splitWord16 nn in
  case r2 of
    M_AF -> RegisterStates hi b c d e (lo.&.0xF0) h l pc sp
    M_BC -> RegisterStates a hi lo d e f h l pc sp
    M_DE -> RegisterStates a b c hi lo f h l pc sp
    M_HL -> RegisterStates a b c d e f hi lo pc sp
    M_PC -> RegisterStates a b c d e f h l nn sp
    M_SP -> RegisterStates a b c d e f h l pc nn

initialA_GB, initialA_GBP, initialA_GBC :: Word8
initialA_GB  = 0x01
initialA_GBP = 0xFF
initialA_GBC = 0x11

initialRegisterStates :: RegisterStates
initialRegisterStates = RegisterStates
  initialA_GB  -- A
  0x00         -- B
  0x13         -- C
  0x00         -- D
  0xD8         -- E
  0xB0         -- F
  0x01         -- H
  0x4D         -- L
  0x0100       -- PC
  0xFFFE       -- SP

vBlankPeriod = 70224
hBlankPeriod = 456

divPeriod = 256

data IrqStates = IrqStates {
  irqStateIME :: Bool,               -- Interrupt Master Enable
  irqStateVBlankCounter :: Int,      -- CPU cycles until next V-Blank (mode1)
  irqStateHBlankCounter :: Int,      -- CPU cycles until next H-Blank (mode2)
  irqStateHBlankMode3Counter :: Int, -- CPU cycles until next H-Blank mode3 cycle
  irqStateHBlankMode0Counter :: Int, -- CPU cycles until next H-Blank mode0 cycle
  irqStateCurrentScanline :: Int,    -- Current scanline, 0-153
  irqStateDisplay :: Display,        -- LCD Display pixels
  irqStateVBlankNow :: Bool,         -- VBlank happened right now at the current instruction
  irqStateDIVCounter :: Int          -- DIV register
}

initialIrqStates = IrqStates {
  irqStateIME = False,
  irqStateVBlankCounter = 0,
  irqStateHBlankCounter = 0,
  irqStateHBlankMode3Counter = 80,
  irqStateHBlankMode0Counter = (80 + 172),
  irqStateCurrentScanline = 153,
  irqStateDisplay = blankDisplay,
  irqStateVBlankNow = False,
  irqStateDIVCounter = 0
}

irqUpdate :: CycleCount ->
             Maybe Bool ->
             ((RegisterStates, Memory), IrqStates) ->
             ((RegisterStates, Memory), IrqStates)
irqUpdate cycles ime = execState $ do
  when (isJust ime) (modify $ transformIrq (\i -> i { irqStateIME = fromJust ime }))
  let updateCounters i = i { irqStateVBlankCounter = (irqStateVBlankCounter i) - cycles,
                             irqStateHBlankCounter = (irqStateHBlankCounter i) - cycles,
                             irqStateHBlankMode3Counter = (irqStateHBlankMode3Counter i) - cycles,
                             irqStateHBlankMode0Counter = (irqStateHBlankMode0Counter i) - cycles,
                             irqStateDIVCounter = (irqStateDIVCounter i) - cycles }
  modify $ transformIrq updateCounters

  modify $ transformIrq (\i -> i { irqStateVBlankNow = (irqStateVBlankCounter i) <= 0 })

  (_, irq) <- get
  when ((irqStateHBlankCounter irq) <= 0)
       (do modify $ transformIrq (\i -> i { irqStateHBlankCounter = (irqStateHBlankCounter i) + hBlankPeriod })
           modify tickHBlank)
  when ((irqStateHBlankMode3Counter irq) <= 0)
       (do modify $ transformIrq (\i -> i { irqStateHBlankMode3Counter = (irqStateHBlankMode3Counter i) + hBlankPeriod })
           modify tickHBlankMode3)
  when ((irqStateHBlankMode0Counter irq) <= 0)
       (do modify $ transformIrq (\i -> i { irqStateHBlankMode0Counter = (irqStateHBlankMode0Counter i) + hBlankPeriod })
           modify tickHBlankMode0)
  when ((irqStateVBlankCounter irq) <= 0)
       (do modify $ transformIrq (\i -> i { irqStateVBlankCounter = (irqStateVBlankCounter i) + vBlankPeriod })
           modify tickVBlank)
  when ((irqStateDIVCounter irq) <= 0)
       (do modify $ transformIrq (\i -> i { irqStateDIVCounter = (irqStateDIVCounter i) + divPeriod })
           modify tickDIV)

  ((_, mem), irq2) <- get
  let ime = irqStateIME irq2
      flagsIF = (memRam mem)!0xFF0F
      flagsIE = (memRam mem)!0xFFFF
  when (ime && (flagsIF .&. flagsIE > 0))
       (let i = getLowBit (flagsIF .&. flagsIE)
            jumpAddr = case i of
                         0 -> 0x0040
                         1 -> 0x0048
                         2 -> 0x0050
                         3 -> 0x0058
                         4 -> 0x0060 in
        do modify $ transformIrq (\i -> i { irqStateIME = False })
           modify $ transformMem (transformMemoryAddr (`clearBit` i) 0xFF0F)
           ((reg, mem), _) <- get
           let oldPC = getReg2State reg M_PC
               (hiPC, loPC) = splitWord16 oldPC
               oldSP = getReg2State reg M_SP
           (flip const) (showHex2 oldSP) (modify $ transformMem ( \m -> writeMem (writeMem m (oldSP-1) hiPC) (oldSP-2) loPC ))
           modify $ transformReg (\r -> setReg2State r M_SP $! (oldSP-2))
           modify $ transformReg (\r -> setReg2State r M_PC $! jumpAddr)
           return ())

  where getLowBit :: Word8 -> Int
        getLowBit n = fromJust (elemIndex True (map (testBit n) [0..4]))

transformReg :: (RegisterStates -> RegisterStates) ->
                ((RegisterStates, Memory), IrqStates) -> 
                ((RegisterStates, Memory), IrqStates)
transformReg t ((r, m), i) = ((t r, m), i)

transformMem :: (Memory -> Memory) ->
                ((RegisterStates, Memory), IrqStates) ->
                ((RegisterStates, Memory), IrqStates)
transformMem t ((r, m), i) = ((r, t m), i)

transformIrq :: (IrqStates -> IrqStates) ->
                ((RegisterStates, Memory), IrqStates) ->
                ((RegisterStates, Memory), IrqStates)
transformIrq t ((r, m), i) = ((r, m), t i)

transformMemoryAddr :: (Word8 -> Word8) -> Word16 -> Memory -> Memory
transformMemoryAddr t a m = m { memRam = (memRam m)//[(a, (t ((memRam m)!a)))] }

tickHBlank :: ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
tickHBlank = execState $ do
  let incrementLY = transformMemoryAddr ((`mod` 154).(+1)) 0xFF44
  modify $ transformMem incrementLY
  modify $ transformIrq (\i -> i { irqStateCurrentScanline = (((`mod` 154).(+1)) (irqStateCurrentScanline i)) } )
  ((_, _), irq) <- get
  when ((irqStateCurrentScanline irq) < 144) $ do
    ((_, mem), _) <- get
    let (ly, lyc) = ((memRam mem)!0xFF44, (memRam mem)!0xFF45)
        (statB5, statB6) = ((memRam mem)!0xFF41 `testBit` 5, (memRam mem)!0xFF41 `testBit` 6)
        updateSTATFlags = transformMemoryAddr
                            (execState $ do
                               modify (`clearBit` 0)
                               modify (`setBit` 1)
                               modify (if ly==lyc then (`setBit` 2) else (`clearBit` 2)))
                            (0xFF41)
        updateIF = transformMemoryAddr
                     (execState $ when (statB5 || (statB6 && ly==lyc)) (modify (`setBit` 1)))
                     (0xFF0F)
    modify $ transformMem updateSTATFlags
    modify $ transformMem updateIF
    modify $ renderScanLine (irqStateCurrentScanline irq)

tickHBlankMode3 :: ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
tickHBlankMode3 = execState $ do
  ((_, _), irq) <- get
  when ((irqStateCurrentScanline irq) < 144) $ do
    let updateSTATFlags = transformMemoryAddr
                            (execState $ do
                               modify (`setBit` 0)
                               modify (`setBit` 1))
                            (0xFF41)
    modify $ transformMem updateSTATFlags

tickHBlankMode0 :: ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
tickHBlankMode0 = execState $ do
  ((_, _), irq) <- get
  when ((irqStateCurrentScanline irq) < 144) $ do
    ((_, mem), _) <- get
    let statB3 = (memRam mem)!0xFF41 `testBit` 3
        updateSTATFlags = transformMemoryAddr
                            (execState $ do
                               modify (`clearBit` 0)
                               modify (`clearBit` 1))
                            (0xFF41)
        updateIF = transformMemoryAddr
                     (execState $ when statB3 (modify (`setBit` 1)))
                     (0xFF0F)
    modify $ transformMem updateSTATFlags
    modify $ transformMem updateIF

tickVBlank :: ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
tickVBlank = execState $ do
  ((_, mem), _) <- get
  let statB4 = (memRam mem)!0xFF41 `testBit` 4
      updateSTATFlags = transformMemoryAddr
                          (execState $ do
                             modify (`setBit` 0)
                             modify (`clearBit` 1))
                          (0xFF41)
      updateIF = transformMemoryAddr
                   (execState $ do
                      when statB4 (modify (`setBit` 1))
                      modify (`setBit` 0))
                   (0xFF0F)
  modify $ transformMem updateSTATFlags
  modify $ transformMem updateIF

tickDIV :: ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
tickDIV = execState $ do
  ((_, mem), _) <- get
  modify $ transformMem (transformMemoryAddr (+1) 0xFF04)

renderScanLine :: Int -> ((RegisterStates, Memory), IrqStates) -> ((RegisterStates, Memory), IrqStates)
renderScanLine scanline = execState $ do
  ((_, mem), irq) <- get
  let d = irqStateDisplay irq
  let scx = fromIntegral (readMem mem 0xFF43)
  let scy = fromIntegral (readMem mem 0xFF42)
  let lcdc = readMem mem 0xFF40
  let lcdon = testBit lcdc 7
  let bgon = testBit lcdc 0
  let bgmap = testBit lcdc 3
  let bgmapStartAddr = if bgmap then 0x9C00 else 0x9800
  let bgtiles = testBit lcdc 4
  let bgtilesStartAddr = if bgtiles then 0x8000 else 0x9000
  let spritesOn = testBit lcdc 1
  let spritesBig = testBit lcdc 2
  let getBgPixel :: Int -> Int -> Word8
      getBgPixel x y =
        let x' = (x + scx) `mod` 256
            y' = (y + scy) `mod` 256
            yrow = y' `div` 8
            xrow = x' `div` 8
            tileNum = yrow * 32 + xrow
            tileIndex = (fromIntegral (readMem mem ((fromIntegral tileNum) + bgmapStartAddr)))::Int8
            tileStartMem = bgtilesStartAddr + (16 * fromIntegral tileIndex)
            xoff = 7 - (x' `mod` 8)
            yoff = y' `mod` 8
            hiByte = tileStartMem + (yoff * 2)
            loByte = tileStartMem + (yoff * 2) + 1
            hiByteValue = readMem mem (fromIntegral hiByte)
            loByteValue = readMem mem (fromIntegral loByte)
            color = (2 * (fromEnum (testBit loByteValue xoff))) + (fromEnum (testBit hiByteValue xoff))
        in if bgon && lcdon then fromIntegral color else 0
  let getSpritesPixel :: Int -> Int -> Maybe Word8
      getSpritesPixel x y = case spritesBig of
        False -> let xposes = map ( \i -> readMem mem (0xFE00 + (i*4) + 1) ) [0..39]
                     yposes = map ( \i -> readMem mem (0xFE00 + (i*4) + 0) ) [0..39]
                     tileIndexes = map ( \i -> readMem mem (0xFE00 + (i*4) + 2) ) [0..39]
                     xflips = map ( \i -> testBit (readMem mem (0xFE00 + (i*4) + 2)) 5 ) [0..39]
                     yflips = map ( \i -> testBit (readMem mem (0xFE00 + (i*4) + 2)) 4 ) [0..39]
                     getSingleSpritePixel :: Int -> Int -> Int -> Int -> Word8 -> Bool -> Bool -> Maybe Word8
                     getSingleSpritePixel x y xpos ypos tileIndex xflip yflip =
                       if x > (fromIntegral xpos)-8 && x <= xpos &&
                          y > (fromIntegral ypos)-16 && y <= ypos-8
                       then let x' = x - (fromIntegral xpos) + 8
                                y' = y - (fromIntegral ypos) + 16
                                x'' = if xflip then 7-x' else x'
                                y'' = if yflip then 7-y' else y'
                                tileStartMem = 0x8000 + (16 * (fromIntegral tileIndex))
                                xoff = 7 - (x'' `mod` 8)
                                yoff = y'' `mod` 8
                                hiByte = tileStartMem + (yoff * 2)
                                loByte = tileStartMem + (yoff * 2) + 1
                                hiByteValue = readMem mem (fromIntegral hiByte)
                                loByteValue = readMem mem (fromIntegral loByte)
                                color = (2 * (fromEnum (testBit loByteValue xoff))) + (fromEnum (testBit hiByteValue xoff))
                            in if color > 0 then Just (fromIntegral color) else Nothing
                       else Nothing
                 in foldl' ( \a b -> if isJust a
                                     then a
                                     else getSingleSpritePixel x
                                                               y
                                                               (fromIntegral (xposes!!b))
                                                               (fromIntegral (yposes!!b))
                                                               (tileIndexes!!b)
                                                               (xflips!!b)
                                                               (yflips!!b)
                           ) Nothing [0..39]
        True -> trace ("eek") Nothing

  let getPixel x y = case getSpritesPixel x y of
                       Just c -> c
                       Nothing -> getBgPixel x y

  let updateList = map (\x -> ((scanline, x), getBgPixel x scanline)) [0..159]
  let d' = d//updateList
  modify $ transformIrq (\i -> i { irqStateDisplay = d' })


machineCpuExecute :: (MemoryModel m) =>
                     (RegisterStates, m) ->
                     ExecutionAST () ->
                     (RegisterStates, m)
machineCpuExecute s e = fst (machineCpuExecute' s e)

machineCpuExecute' :: (MemoryModel m) =>
                      (RegisterStates, m) ->
                      ExecutionAST a ->
                      ((RegisterStates, m), a)
machineCpuExecute' state@(regS, memS) e = case e of
  Return result -> (state, result)
  Bind l r -> let (s, result) = machineCpuExecute' state l in
              machineCpuExecute' s (r result)
  WriteRegister reg n -> ((setRegState regS reg n, memS), ())
  ReadRegister reg -> (state, getRegState regS reg)
  WriteRegister2 reg2 nn -> ((setReg2State regS reg2 nn, memS), ())
  ReadRegister2 reg2 -> (state, getReg2State regS reg2)
  WriteMemory a n -> ((regS, writeMem memS a n), ())
  ReadMemory a -> (state, readMem memS a)

fetchInstruction :: (MemoryModel m) => (RegisterStates, m) -> Instruction
fetchInstruction (regS, memS) =
  let pc = getReg2State regS M_PC
      opcode = readMem memS pc
      n :: Word8
      n = readMem memS (pc+1)
      nn :: Word16
      nn = joinWord16 (readMem memS (pc+2)) (readMem memS (pc+1))
      instruction = machineCodeToInstruction opcode (n, nn) in
  instruction

machineStepInstruction :: (MemoryModel m) =>
                          (RegisterStates, m) ->
                          (RegisterStates, m)
machineStepInstruction state@(regS, memS) =
  let instruction = fetchInstruction state
      execution = executeInstruction instruction in
  machineCpuExecute state execution

updateMachine :: ((RegisterStates, Memory), IrqStates) ->
                 ((RegisterStates, Memory), IrqStates)
updateMachine (state@(regS, memS), irqS) =
  let stepInstruction = machineStepInstruction state
      pc = getReg2State regS M_PC
      opcode = readMem memS pc
      cycles = opcodeCycleCount opcode
      ime = opcodeQueryIME opcode in
  irqUpdate cycles ime (stepInstruction, irqS)

initialMachineState romImage =
  ((initialRegisterStates, initMemory romImage), initialIrqStates)

--instance Show JoypadKeyStates

updateMachineDisplayFrame :: JoypadKeyStates ->
                             ((RegisterStates, Memory), IrqStates) ->
                             (Display, ((RegisterStates, Memory), IrqStates))
updateMachineDisplayFrame jp s =
  let s' = transformMem (\mem -> mem { memJoypadKeyStates = jp } ) s
      l = tail (iterate updateMachine s')
      pred (_, irq) = irqStateVBlankNow irq
      f@(_, irqS) = fromJust (find pred l) in
  (irqStateDisplay irqS, f)

--updateMachineDisplayFrame' jp s =
--  let (_, s') = updateMachineDisplayFrame jp s in
--  updateMachineDisplayFrame jp s'

